File: script/commands/toggle.py
"""
toggle 指令：啟用/停用特定工具的特定資源。
"""

from typing import Optional

import typer
from rich.console import Console
from rich.table import Table

from ..utils.shared import (
    load_toggle_config,
    disable_resource,
    enable_resource,
)

app = typer.Typer()
console = Console()

TARGET_DISPLAY_NAMES = {
    "claude": "Claude Code",
    "antigravity": "Antigravity",
    "opencode": "OpenCode",
    "codex": "Codex",
    "gemini": "Gemini CLI",
}


@app.command()
def toggle(
    target: Optional[str] = typer.Option(
        None,
        "--target",
        "-t",
        help="目標工具：claude, antigravity, opencode, codex, gemini",
    ),
    resource_type: Optional[str] = typer.Option(
        None,
        "--type",
        "-T",
        help="資源類型：skills, commands, agents, workflows",
    ),
    name: Optional[str] = typer.Option(
        None,
        "--name",
        "-n",
        help="資源名稱",
    ),
    enable: bool = typer.Option(
        False,
        "--enable",
        "-e",
        help="啟用資源",
    ),
    disable: bool = typer.Option(
        False,
        "--disable",
        "-d",
        help="停用資源",
    ),
    list_status: bool = typer.Option(
        False,
        "--list",
        "-l",
        help="列出目前的開關狀態",
    ),
):
    """啟用/停用特定工具的特定資源。"""
    config = load_toggle_config()

    # 列出狀態
    if list_status:
        show_toggle_status(config)
        return

    # 驗證參數
    if not target:
        console.print("[red]請指定目標工具（--target）[/red]")
        raise typer.Exit(code=1)

    valid_targets = ["claude", "antigravity", "opencode", "codex", "gemini"]
    if target not in valid_targets:
        console.print(f"[red]無效的目標：{target}[/red]")
        raise typer.Exit(code=1)

    if not resource_type:
        console.print("[red]請指定資源類型（--type）[/red]")
        raise typer.Exit(code=1)

    valid_types_for_target = {
        "claude": ["skills", "commands"],
        "antigravity": ["skills", "workflows"],
        "opencode": ["agents"],
        "codex": ["skills"],
        "gemini": ["skills", "commands"],
    }

    if resource_type not in valid_types_for_target.get(target, []):
        console.print(f"[red]{target} 不支援 {resource_type} 類型[/red]")
        console.print(f"有效選項：{', '.join(valid_types_for_target[target])}")
        raise typer.Exit(code=1)

    if not name:
        console.print("[red]請指定資源名稱（--name）[/red]")
        raise typer.Exit(code=1)

    if enable == disable:
        console.print("[red]請指定 --enable 或 --disable（二擇一）[/red]")
        raise typer.Exit(code=1)

    # 確保結構存在
    if target not in config:
        config[target] = {}
    if resource_type not in config[target]:
        config[target][resource_type] = {"enabled": True, "disabled": []}

    disabled_list = config[target][resource_type].get("disabled", [])

    if disable:
        if name in disabled_list:
            console.print(f"[dim]{name} 已經是停用狀態[/dim]")
        else:
            # 使用檔案移動機制停用資源
            if not disable_resource(target, resource_type, name):
                raise typer.Exit(code=1)
    else:  # enable
        if name not in disabled_list:
            console.print(f"[dim]{name} 已經是啟用狀態[/dim]")
        else:
            # 使用檔案還原機制啟用資源
            if not enable_resource(target, resource_type, name):
                raise typer.Exit(code=1)


def show_toggle_status(config: dict):
    """顯示目前的開關狀態。"""
    table = Table(title="Toggle 狀態")
    table.add_column("目標", style="cyan")
    table.add_column("類型", style="green")
    table.add_column("整體啟用", style="yellow")
    table.add_column("停用項目", style="red")

    for target, types in config.items():
        target_name = TARGET_DISPLAY_NAMES.get(target, target)
        for resource_type, settings in types.items():
            enabled = settings.get("enabled", True)
            disabled = settings.get("disabled", [])

            enabled_text = "✓" if enabled else "✗"
            disabled_text = ", ".join(disabled) if disabled else "-"

            table.add_row(target_name, resource_type, enabled_text, disabled_text)

    console.print(table)

File: src/cli/toggle.ts
import { access, readdir, rename } from "node:fs/promises";
import { join } from "node:path";

import type { Command } from "commander";
import chalk from "chalk";

import { printError, printTable } from "../utils/formatter";
import { t } from "../utils/i18n";
import {
  COPY_TARGETS,
  type ResourceType,
  TARGET_NAMES,
  type TargetType,
} from "../utils/shared";

const DISABLED_SUFFIX = ".disabled";

type ToggleStateItem = {
  target: TargetType;
  type: ResourceType;
  name: string;
  enabled: boolean;
};

async function pathExists(path: string): Promise<boolean> {
  try {
    await access(path);
    return true;
  } catch {
    return false;
  }
}

function isTarget(value: string): value is TargetType {
  return value in COPY_TARGETS;
}

function isResourceType(value: string): value is ResourceType {
  return (
    value === "skills" ||
    value === "commands" ||
    value === "agents" ||
    value === "workflows"
  );
}

function validateTargetType(
  target: string | undefined,
  type: string | undefined,
):
  | { ok: true; target: TargetType; type: ResourceType; basePath: string }
  | { ok: false; message: string } {
  if (!target || !type) {
    return { ok: false, message: t("toggle.missing_target_type") };
  }

  if (!isTarget(target)) {
    return { ok: false, message: t("toggle.invalid_target", { target }) };
  }

  if (!isResourceType(type)) {
    return { ok: false, message: t("toggle.invalid_type", { type }) };
  }

  const basePath = COPY_TARGETS[target][type];
  if (!basePath) {
    return {
      ok: false,
      message: t("toggle.invalid_combo", { target, type }),
    };
  }

  return { ok: true, target, type, basePath };
}

function resourceFileName(type: ResourceType, name: string): string {
  if (type === "skills") {
    return name;
  }
  return name.endsWith(".md") ? name : `${name}.md`;
}

async function disableResource(
  target: TargetType,
  type: ResourceType,
  name: string,
): Promise<boolean> {
  const base = COPY_TARGETS[target][type];
  if (!base) {
    return false;
  }

  const activePath = join(base, resourceFileName(type, name));
  const disabledPath = `${activePath}${DISABLED_SUFFIX}`;

  if (!(await pathExists(activePath))) {
    return false;
  }

  await rename(activePath, disabledPath);
  return true;
}

async function enableResource(
  target: TargetType,
  type: ResourceType,
  name: string,
): Promise<boolean> {
  const base = COPY_TARGETS[target][type];
  if (!base) {
    return false;
  }

  const activePath = join(base, resourceFileName(type, name));
  const disabledPath = `${activePath}${DISABLED_SUFFIX}`;

  if (!(await pathExists(disabledPath))) {
    return false;
  }

  await rename(disabledPath, activePath);
  return true;
}

function parseToggleEntry(
  type: ResourceType,
  rawName: string,
): ToggleStateItem["name"] {
  const withoutDisabled = rawName.endsWith(DISABLED_SUFFIX)
    ? rawName.slice(0, -DISABLED_SUFFIX.length)
    : rawName;

  if (type === "skills") {
    return withoutDisabled;
  }

  return withoutDisabled.endsWith(".md")
    ? withoutDisabled.slice(0, -3)
    : withoutDisabled;
}

async function listToggleState(
  target: TargetType,
  type: ResourceType,
): Promise<ToggleStateItem[]> {
  const base = COPY_TARGETS[target][type];
  if (!base) {
    return [];
  }

  try {
    const entries = await readdir(base, { withFileTypes: true });
    return entries
      .filter((entry) => {
        if (entry.name.startsWith(".")) {
          return false;
        }
        return type === "skills" ? entry.isDirectory() : entry.isFile();
      })
      .map((entry) => ({
        target,
        type,
        name: parseToggleEntry(type, entry.name),
        enabled: !entry.name.endsWith(DISABLED_SUFFIX),
      }))
      .sort((a, b) => a.name.localeCompare(b.name));
  } catch {
    return [];
  }
}

async function listAllToggleStates(
  target?: string,
  type?: string,
): Promise<ToggleStateItem[] | { error: string }> {
  if (target || type) {
    const validated = validateTargetType(target, type);
    if (!validated.ok) {
      return { error: validated.message };
    }
    return await listToggleState(validated.target, validated.type);
  }

  const items: ToggleStateItem[] = [];

  for (const targetKey of Object.keys(COPY_TARGETS) as TargetType[]) {
    for (const typeKey of [
      "skills",
      "commands",
      "agents",
      "workflows",
    ] as const) {
      if (!COPY_TARGETS[targetKey][typeKey]) {
        continue;
      }
      items.push(...(await listToggleState(targetKey, typeKey)));
    }
  }

  return items.sort((a, b) => {
    const byTarget = a.target.localeCompare(b.target);
    if (byTarget !== 0) {
      return byTarget;
    }
    const byType = a.type.localeCompare(b.type);
    if (byType !== 0) {
      return byType;
    }
    return a.name.localeCompare(b.name);
  });
}

export function registerToggleCommand(program: Command): void {
  program
    .command("toggle")
    .description(t("cmd.toggle"))
    .option("-t, --target <target>", t("opt.target"))
    .option("-T, --type <type>", t("opt.type"))
    .option("-n, --name <name>", t("opt.name"))
    .option("-e, --enable", t("opt.enable"))
    .option("-d, --disable", t("opt.disable"))
    .option("-l, --list", t("opt.list"))
    .option("--all", t("opt.toggle_all"))
    .action(
      async (options: {
        target?: string;
        type?: string;
        name?: string;
        enable?: boolean;
        disable?: boolean;
        list?: boolean;
        all?: boolean;
      }) => {
        if (options.list) {
          const listed = await listAllToggleStates(
            options.target,
            options.type,
          );
          if (!Array.isArray(listed)) {
            printError(listed.error);
            process.exitCode = 1;
            return;
          }

          if (listed.length === 0) {
            console.log(t("list.no_resources"));
            return;
          }

          printTable(
            ["Target", "Type", "Name", "Status"],
            listed.map((item) => [
              item.target,
              item.type,
              item.name,
              item.enabled ? "Enabled" : "Disabled",
            ]),
          );
          return;
        }

        const validated = validateTargetType(options.target, options.type);
        if (!validated.ok) {
          printError(validated.message);
          process.exitCode = 1;
          return;
        }

        const targetName = TARGET_NAMES[validated.target] ?? validated.target;

        // --all 批次模式
        if (options.all && !options.name) {
          if (options.enable === options.disable) {
            printError(t("toggle.choose_one"));
            process.exitCode = 1;
            return;
          }

          const resources = await listToggleState(validated.target, validated.type);
          let count = 0;
          for (const r of resources) {
            if (options.enable && !r.enabled) {
              const ok = await enableResource(validated.target, validated.type, r.name);
              if (ok) count++;
            } else if (options.disable && r.enabled) {
              const ok = await disableResource(validated.target, validated.type, r.name);
              if (ok) count++;
            }
          }

          const action = options.enable ? t("common.enabled") : t("common.disabled");
          console.log(chalk.green(`✓ ${t("toggle.batch_done", { action, count: String(count), target: targetName, type: validated.type })}`));
          console.log(chalk.dim(t("toggle.restart_hint", { target: targetName })));
          return;
        }

        // 單一資源模式
        if (!options.name) {
          printError(t("toggle.missing_name"));
          process.exitCode = 1;
          return;
        }

        if (options.enable === options.disable) {
          printError(t("toggle.choose_one"));
          process.exitCode = 1;
          return;
        }

        const success = options.enable
          ? await enableResource(validated.target, validated.type, options.name)
          : await disableResource(validated.target, validated.type, options.name);

        if (!success) {
          // 可能是已經在目標狀態
          if (options.enable) {
            console.log(chalk.dim(t("toggle.already_enabled", { name: options.name })));
          } else {
            console.log(chalk.dim(t("toggle.already_disabled", { name: options.name })));
          }
          return;
        }

        if (options.enable) {
          console.log(chalk.green(`✓ ${t("toggle.enabled", { target: targetName, type: validated.type, name: options.name })}`));
        } else {
          console.log(chalk.yellow(`✓ ${t("toggle.disabled", { target: targetName, type: validated.type, name: options.name })}`));
        }
        console.log(chalk.dim(t("toggle.restart_hint", { target: targetName })));
      },
    );
}
